Backend-tests.md

```diff
@@ -0,0 +1,596 @@
+
+   Also, try writing contrasting tests. For example, if you are checking that an exception is raised under a certain criterion, also add a test to ensure that the exception is not raised when the criterion is not satisfied.
+
+7. For **test outputs**, follow these guidelines:
+
+   * Test each output as exactly and completely as possible. For example, it's better to compare equality for an entire dict rather than just checking that a particular value has changed.
+   * Use `assertTrue(value)` or `assertFalse(value)` instead of `assertEqual(value, True)` or `assertEqual(value, False)`.
+   * Use `assertIsNone(value)` instead of `assertEqual(value, None)`.
+
+8. If you create a new test module (a `*_test.py` file), you will need to add it to a shard in [`oppia/scripts/backend_test_shards.json`](https://github.com/oppia/oppia/blob/develop/scripts/backend_test_shards.json). See the [section on shards below](#sharding-backend-tests).
+
+### Sharding backend tests
+
+#### How sharding works
+
+We shard the backend tests into many smaller jobs that run in parallel on GitHub Actions. In your PRs, you'll see something like this:
+
+![Display of backend test jobs on a pull request](https://user-images.githubusercontent.com/19878639/109242853-ddd78700-77a9-11eb-9cae-da5cece9ab26.png)
+
+The jobs named like `Run backend tests (ubuntu-18.04, i)` are running the sharded jobs, each with its own number `i`. The last job, `Check combined backend test coverage`, checks that the code coverage is 100%. All these jobs are defined in [`.github/workflows/backend_tests.yml`](https://github.com/oppia/oppia/blob/develop/.github/workflows/backend_tests.yml).
+
+The shards are defined in [`scripts/backend_test_shards.json`](https://github.com/oppia/oppia/blob/develop/scripts/backend_test_shards.json). Here's what the top of that file looks like:
+
+```json
+{
+    "1": [
+        "core.controllers.base_test",
+        "core.controllers.collection_editor_test",
+```
+
+Each shard is identified by a name, in this case `1`. You could run this shard with `python -m scripts.run_backend_tests --test_shard 1`. The shards are then defined by a list of test modules. For example, the module name for `core/controllers/base_test.py` is `core.controllers.base_test`. Notice that there is no `.py` at the end!
+
+#### Common errors
+
+Whenever you run a shard of backend tests, the `run_backend_tests.py` script checks to make sure the test modules on the filesystem and the modules in the shards file are exactly the same. If a shard has a module that isn't in the filesystem, you'll get a `Modules ... in shards not found` error. This often happens if the module name is incorrect. On the other hand, if there is a module on the filesystem that's not in the shards, you'll get a `Modules ... not in shards` error. This often happens because you forgot to add a new test to the shards.
```

Comment:
I'm having difficulty parsing "Modules ... in shards not found". Perhaps this can be worded better?

Not clear whether it's the modules or the shards that are not found.

Also perhaps it might be nice to add an addendum to the error message raised telling the developer what to do? E.g. for the "Modules ... not in shards" the error message can say "make sure that X is added to one of the shards in the file Y."

Reply:
I've opened oppia/oppia#13824 to track fixing these errors